### ISSUE TITLE
vmalert data loss

### ISSUE THREAD

**Issue by Dennis8274** [VictoriaMetrics/VictoriaMetrics#5661]
### Describe the bug

found some log as follows:
[WARN] attempt 1 to send request failed: error while sending request to http://xxxx:8480/insert/0/prometheus/api/v1/write: Post "http://xxxx:8480/insert/0/prometheus/api/v1/write": EOF; Data len 233(233) (retriable: true)
[ERRO] attempts to send remote-write request failed - dropping 1 time series
WARN] attempt 1 to send request failed: error while sending request to http://xxxx:8480/insert/0/prometheus/api/v1/write: Post "http://xxxx:8480/insert/0/prometheus/api/v1/write": read tcp xxxx:58204->xxxx:8480: read: connection reset by peer; Data len 233(233) (retriable: true)
ERRO] attempts to send remote-write request failed - dropping 1 time series

time series data loss? pls

### To Reproduce

start vmalert with remoteWrite.url flag

### Version

vminsert-20240123-040413-heads-cluster-0-gf301dc5cf

**Comment by dmitryk-dk** [comment]
Hi @Dennis8274 ! 
vmalert execute queries against remote datasource which has reliability risks because of the network. It is recommended to configure alerts thresholds and rules expressions with the understanding that network requests may fail;

**Comment by Dennis8274** [comment]
> Hi @Dennis8274 ! vmalert execute queries against remote datasource which has reliability risks because of the network. It is recommended to configure alerts thresholds and rules expressions with the understanding that network requests may fail;
> 
> [docs.victoriametrics.com/vmalert.html#limitations](https://docs.victoriametrics.com/vmalert.html#limitations)

thanks for your reply.
em...,i already got all you have said, but i still suggest that vmalert should retry in case of connection reset/disconnected by remote side. In the scenario of vmalert pushing states to vminsert,  the connection periodically(2 minutes default) reset by vminsert, for load balancing of vminsert  consideration.

**Comment by dmitryk-dk** [comment]
> > Hi @Dennis8274 ! vmalert execute queries against remote datasource which has reliability risks because of the network. It is recommended to configure alerts thresholds and rules expressions with the understanding that network requests may fail;
> > [docs.victoriametrics.com/vmalert.html#limitations](https://docs.victoriametrics.com/vmalert.html#limitations)
> 
> thanks for your reply. em...,i already got all you have said, but i still suggest that vmalert should retry in case of connection reset/disconnected by remote side. In the scenario of vmalert pushing states to vminsert, the connection periodically(2 minutes default) reset by vminsert, for load balancing of vminsert consideration.

Hi @Dennis8274 ! Can you check that you are using this flags on the `vmalert` side
```golang
	retryMinInterval  = flag.Duration("remoteWrite.retryMinInterval", time.Second, "The minimum delay between retry attempts. Every next retry attempt will double the delay to prevent hammering of remote database. See also -remoteWrite.retryMaxInterval")
	retryMaxTime      = flag.Duration("remoteWrite.retryMaxTime", time.Second*30, "The max time spent on retry attempts for the failed remote-write request. Change this value if it is expected for remoteWrite.url to be unreachable for more than -remoteWrite.retryMaxTime. See also -remoteWrite.retryMinInterval")
```

And you do not change them

**Comment by thunderbird86** [comment]
I also have the same issue. Have few exemplairs vmalert->vmsigle  for few clusters. This appers only on one vmalert. There was no network issues.
```
{"ts":"2024-04-04T16:24:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:40576->10.243.204.195:8428: read: connection reset by peer; Data len 1151(1151) (retriable: true)"}
{"ts":"2024-04-04T22:17:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:58262->10.243.204.195:8428: read: connection reset by peer; Data len 2581(2581) (retriable: true)"}
{"ts":"2024-04-04T23:34:41.903Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:48742->10.243.204.195:8428: read: connection reset by peer; Data len 1198(1198) (retriable: true)"}
{"ts":"2024-04-05T02:58:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:44176->10.243.204.195:8428: read: connection reset by peer; Data len 2563(2563) (retriable: true)"}
{"ts":"2024-04-05T20:41:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:52004->10.243.204.195:8428: read: connection reset by peer; Data len 2596(2596) (retriable: true)"}
{"ts":"2024-04-05T21:09:41.903Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:38572->10.243.204.195:8428: read: connection reset by peer; Data len 2528(2528) (retriable: true)"}
{"ts":"2024-04-06T01:14:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:56408->10.243.204.195:8428: read: connection reset by peer; Data len 1030(1030) (retriable: true)"}
{"ts":"2024-04-06T02:27:41.903Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:35350->10.243.204.195:8428: read: connection reset by peer; Data len 2533(2533) (retriable: true)"}
{"ts":"2024-04-06T07:13:41.905Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:49788->10.243.204.195:8428: read: connection reset by peer; Data len 1151(1151) (retriable: true)"}
{"ts":"2024-04-06T08:15:41.903Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:51312->10.243.204.195:8428: read: connection reset by peer; Data len 1129(1129) (retriable: true)"}
{"ts":"2024-04-06T11:23:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:42024->10.243.204.195:8428: read: connection reset by peer; Data len 1239(1239) (retriable: true)"}
{"ts":"2024-04-06T13:05:41.903Z","level":"warn","caller":"VictoriaMetrics/app/vmalert```

**Comment by dmitryk-dk** [comment]
> I also have the same issue. Have few exemplairs vmalert->vmsigle for few clusters. This appers only on one vmalert. There was no network issues.
> 
> ```
> {"ts":"2024-04-04T16:24:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:40576->10.243.204.195:8428: read: connection reset by peer; Data len 1151(1151) (retriable: true)"}
> {"ts":"2024-04-04T22:17:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:58262->10.243.204.195:8428: read: connection reset by peer; Data len 2581(2581) (retriable: true)"}
> {"ts":"2024-04-04T23:34:41.903Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:48742->10.243.204.195:8428: read: connection reset by peer; Data len 1198(1198) (retriable: true)"}
> {"ts":"2024-04-05T02:58:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:44176->10.243.204.195:8428: read: connection reset by peer; Data len 2563(2563) (retriable: true)"}
> {"ts":"2024-04-05T20:41:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:52004->10.243.204.195:8428: read: connection reset by peer; Data len 2596(2596) (retriable: true)"}
> {"ts":"2024-04-05T21:09:41.903Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:38572->10.243.204.195:8428: read: connection reset by peer; Data len 2528(2528) (retriable: true)"}
> {"ts":"2024-04-06T01:14:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:56408->10.243.204.195:8428: read: connection reset by peer; Data len 1030(1030) (retriable: true)"}
> {"ts":"2024-04-06T02:27:41.903Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:35350->10.243.204.195:8428: read: connection reset by peer; Data len 2533(2533) (retriable: true)"}
> {"ts":"2024-04-06T07:13:41.905Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:49788->10.243.204.195:8428: read: connection reset by peer; Data len 1151(1151) (retriable: true)"}
> {"ts":"2024-04-06T08:15:41.903Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:51312->10.243.204.195:8428: read: connection reset by peer; Data len 1129(1129) (retriable: true)"}
> {"ts":"2024-04-06T11:23:41.904Z","level":"warn","caller":"VictoriaMetrics/app/vmalert/remotewrite/client.go:240","msg":"attempt 1 to send request failed: error while sending request to http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write: Post \"http://vmsingle-server-0.vmsingle-server.vmsingle:8428/api/v1/write\": read tcp 10.242.1.150:42024->10.243.204.195:8428: read: connection reset by peer; Data len 1239(1239) (retriable: true)"}
> {"ts":"2024-04-06T13:05:41.903Z","level":"warn","caller":"VictoriaMetrics/app/vmalert```
> ```

Hi @thunderbird86 ! Can you share the `vmalert` configuration and `vminsert`? 
What versions of both components?

**Comment by valyala** [comment]
The issue should be addressed in the release [v1.102.0-rc1](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.102.0-rc1). See [this pull request](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/6382) for more details.
